clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -65,9 +69,182 @@ public MemberData(List<TopicPartition> partitions, Optional<Integer> generation)
+
+        // initialize the assignment map with an empty array of size minQuota for all members
+        Map<String, List<TopicPartition>> assignment = new HashMap<>(
+            consumerToOwnedPartitions.keySet().stream().collect(Collectors.toMap(c -> c, c -> new ArrayList<>(minQuota))));
+
+        for (Map.Entry<String, List<TopicPartition>> consumerEntry : consumerToOwnedPartitions.entrySet()) {
+            String consumer = consumerEntry.getKey();
+            List<TopicPartition> ownedPartitions = consumerEntry.getValue();
+
+            if (ownedPartitions.size() < minQuota) {
+                assignment.get(consumer).addAll(ownedPartitions);
+                unassignedPartitions.removeAll(ownedPartitions);
+                unfilledMembers.add(consumer);
+            } else if (ownedPartitions.size() == minQuota) {
+                assignment.get(consumer).addAll(ownedPartitions);
+                unassignedPartitions.removeAll(ownedPartitions);
+                minCapacityMembers.add(consumer);
+            } else {
+                List<TopicPartition> assignmentPartitions = assignment.get(consumer);
+                Iterator<TopicPartition> ownedPartitionsIter = ownedPartitions.iterator();
+                for (int i = 0; i < maxQuota; ++i) {
+                    TopicPartition tp = ownedPartitionsIter.next();
+                    assignmentPartitions.add(tp);
+                    unassignedPartitions.remove(tp);
+                }
+                maxCapacityMembers.add(consumer);
+            }
+        }
+
+        Collections.sort(unfilledMembers);
+        Iterator<TopicPartition> unassignedPartitionsIter = unassignedPartitions.iterator();
+
+        while (!unfilledMembers.isEmpty() && !unassignedPartitions.isEmpty()) {
+            Iterator<String> unfilledConsumerIter = unfilledMembers.iterator();
+
+            while (unfilledConsumerIter.hasNext()) {
+                String consumer = unfilledConsumerIter.next();
+                List<TopicPartition> consumerAssignment = assignment.get(consumer);
+
+                if (unassignedPartitionsIter.hasNext()) {
+                    consumerAssignment.add(unassignedPartitionsIter.next());
+                    unassignedPartitionsIter.remove();
+                } else {
+                    break;
+                }
+
+                if (consumerAssignment.size() == minQuota) {
+                    minCapacityMembers.add(consumer);
+                    unfilledConsumerIter.remove();
+                }
+            }
+        }
+
+        // If we ran out of unassigned partitions before filling all consumers, we need to start stealing partitions
+        // from the over-full consumers at max capacity
+        for (String consumer : unfilledMembers) {
+            List<TopicPartition> consumerAssignment = assignment.get(consumer);
+            int remainingCapacity = minQuota - consumerAssignment.size();
+            while (remainingCapacity > 0) {
```

Comment:
Is it possible that this unfilled consumer has N+1 remaining capacity, while there's only N max consumer only?

Reply:
NVM, I realized it should never happen.